### ISSUE TITLE
Route snapper is picking up irrelevant edges

### ISSUE THREAD

**Issue by dabreegster** [nptscot/npw#19]
Footways and things are currently included. Need to do the `graph` refactor to have a customizable mode, and use it here

**Comment by dabreegster** [comment]
95982c2e4fe64e6369e8df21ab5a80be3839f86b started this. But there's more work to do to initially clean up the graph of possibly editable roads:

- [ ] Edinburgh has lots of pedestrian crossings tagged just as `highway=footway`, like https://www.openstreetmap.org/way/400692851. Ideally we'd prune.
- [ ] Need to prune out disconnected components, like `highway=path` in a golf course, https://www.openstreetmap.org/way/909112247

**Comment by Robinlovelace** [comment]
In case of use/interest, we remove irrelevant ways with the the `get_cycling_network()` function early in the process.

- Function documentation: https://nptscot.github.io/osmactive/reference/get_cycling_network.html
- Regex for highway values we exclude: 

```
    "abandoned|bridleway|bus_guideway|byway|construction|corridor|disused|elevator|emergency|escalator|escap",
    "|far|fixme|gallop|historic|motorway|no|planned|platform|proposed|raceway|rest|road|services|steps|track"
```

Source: https://github.com/nptscot/osmactive/blob/e7372dee246af427781e35e863d40cdbbdefe405/R/osmactive.R#L50C1-L52C1

- Bicycle values we exclude:

```
    "mtb|discouraged|unknown|no"
```

**Comment by Robinlovelace** [comment]
It could be beneficial to align these and I think the function is imperfect, so any issues with the outputs let us know. You can see the results here: https://nptscot.github.io/#/clos/#11.15/55.939/-3.2132

**Comment by Robinlovelace** [comment]
Or ideas for improvement..

**Comment by dabreegster** [comment]
Thanks, I was referring to that filtering. It'll be something to iterate on. For NPW, we're not just trying to reason about existing infrastructure, but also let people quickly select roads to consider improvements. If we filter out too much, then somebody has to draw freehand. What if there's a very poor surface `track` and someone's proposal is to pave it and make a proper rural shared use path?

**Comment by Robinlovelace** [comment]
> What if there's a very poor surface `track` and someone's proposal is to pave it and make a proper rural shared use path?

Oh yes, so there are different filtering requirements for the 'currently cyclable network' data that osmactive gets and the 'potentially cyclable network after interventions' network.

**Comment by dabreegster** [comment]
I'm now trying to make sure NPW and NPT match much better on detecting existing infra, and in fact, pruning out some edges from consideration in the first place. The subtlety is that NPW can modify something currently unsuitable to make it become suitable, so we may need to include more cases upfront, but not assign them an existing infra type.

You can now debug this better in the "Existing network" layer by toggling two options. osmactive's view:
![image](https://github.com/user-attachments/assets/666ef4b7-a603-468e-9889-8a78638a8890)
NPW's currently very buggy view:
![image](https://github.com/user-attachments/assets/ca2d2472-8717-4588-a090-45dccc9d1056)

There's lots of shared footways like https://www.openstreetmap.org/way/933658433 that're only tagged `highway=path`, nothing else. https://github.com/nptscot/osmactive/blob/912b61e6d0b6403f1737989bb22a84ca5ea73af7/R/osmactive.R#L141   osmactive is pretty strict about these, totally discarding footways and paths without explicit bicycle designation and explicitly good surfacing.

So the question for @Robinlovelace and @mvl22 -- should we more or less match osmactive's choices in NPW? It's indeed confusing right now that there are so many paths in the middle of nowhere that you can snap to. In the worst case, if we wind up excluding existing paths that the user might want to snap to and use for making real improvements, then they could fall back to drawing new freehand roads. It might also be simpler to start by being stricter and matching osmactive closely, and then relaxing in gradual ways as we do some testing and discover edge cases we do want to include. And when that happens, maybe osmactive might want to change too.

**Comment by dabreegster** [comment]
Filtering out paths without explicit `bicycle=yes|designated` helped massively. Just as a code walkthrough, the approach in NPW is two-stage. https://github.com/nptscot/npw/blob/main/backend/src/existing.rs
First `Highway::classify` decides if an edge should be included in the model at all, and if so, captures the OSM highway type as an explicit enum list: https://github.com/nptscot/npw/blob/569d84372bebc0df17fd20c520eb3848b49eead7/backend/src/existing.rs#L13C1-L25C10

Then to decide if there's any existing infra on something, https://github.com/nptscot/npw/blob/569d84372bebc0df17fd20c520eb3848b49eead7/backend/src/existing.rs#L141 handles things based on that highway enum. I found the osmactive https://github.com/nptscot/osmactive/blob/912b61e6d0b6403f1737989bb22a84ca5ea73af7/R/osmactive.R#L244 equivalent hard to follow, because the cases for roads vs cycleway/path/footway/pedestrian are so different. The results don't match osmactive well yet, but I'm working through the mismatches

**Comment by Robinlovelace** [comment]
> should we more or less match osmactive's choices in NPW?

It's a reasonable starting point but, as you say, it will likely make sense for the NPW editor to have access to more paths. That can be added later though. Great that removing paths *without* certain tags worked. I wouldn't worry too much about `osmactive` compatibility for now as they are quite different use cases. `osmactive` focusses on visualisation existing active travel infrastructure, while the NPW is more about what it could be in the future.

**Comment by mvl22** [comment]
> Edinburgh has lots of pedestrian crossings tagged just as highway=footway, like https://www.openstreetmap.org/way/400692851. Ideally we'd prune.

Personally I would advise against pruning, as inevitably sometimes `highway=footway` will be part of a genuine overall route. For instance, there may be a very short spur going over a footway which provides vital connectivity.

In general I would take a relatively strict approach, as laxity only promotes sloppy mapping. In my view it is generally best to adopt as pure an approach in data consumption as possible, and let errors be visible in the mapping, so as to motivate fixes. OSM data fixes don't usually take that long.

In this example I've spent five minutes fixing up a big pile of those crossings in that area:
https://www.openstreetmap.org/changeset/160306653

If you have mapping showing these, let me know and I'll zap more if possible.

**Comment by mvl22** [comment]
> Need to prune out disconnected components, like highway=path in a golf course, https://www.openstreetmap.org/way/909112247

Yes, disconnected components are very likely to be either stuff like courtyards or paths in the middle of grassed areas, e.g. a ring round a fountain. They can be safety pruned. If something is genuinely part of a path network it will or should be connected.

**Comment by dabreegster** [comment]
I think this is mostly fixed now, closing out old issues
